File: app/src/main/java/pl/pola_app/ui/adapter/ProductsAdapter.java
package pl.pola_app.ui.adapter;

import android.content.Context;
import android.content.res.Resources;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import pl.pola_app.R;
import pl.pola_app.databinding.ViewProductItemBinding;
import pl.pola_app.model.SearchResult;
import timber.log.Timber;

public class ProductsAdapter extends RecyclerView.Adapter<ProductsAdapter.ViewHolder> {

    public interface ProductClickListener {
        void itemClicked(SearchResult searchResult);
    }

    @NonNull
    private final Context context;
    @NonNull
    private final ProductList searchResults;
    @Nullable
    private ProductClickListener productClickListener;

    public ProductsAdapter(@NonNull final Context context,
            @NonNull final ProductList searchResults) {
        this.searchResults = searchResults;
        this.context = context;
    }

    public void setOnProductClickListener(ProductClickListener productClickListener) {
        this.productClickListener = productClickListener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup viewGroup, int i) {
        LayoutInflater inflater = LayoutInflater.from(viewGroup.getContext());
        ViewProductItemBinding binding = ViewProductItemBinding.inflate(inflater, viewGroup, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(final ViewHolder viewHolder, final int i) {
        final SearchResult p = searchResults.get(i);
        viewHolder.bind(p);
    }

    @Override
    public int getItemCount() {
        return searchResults.size();
    }

    private void onItemClicked(int position) {
        if (productClickListener == null) {
            return;
        }

        if (position > searchResults.size()) { //TODO: is it even possible to reach this state?
            Timber.w(new IndexOutOfBoundsException(), "Position: %d, list size: %d", position, searchResults.size());
            return;
        }

        SearchResult searchResult = searchResults.get(position);
        if (searchResult != null) {
            productClickListener.itemClicked(searchResult);
        }
    }

    class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        ViewProductItemBinding binding;

        ViewHolder(ViewProductItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        void bind(SearchResult searchResult) {
            itemView.setOnClickListener(this);
            if (searchResult == null) {
                progress.setVisibility(View.VISIBLE);
                companyName.setText("");
                plScore.setProgress(0);
                binding.progressBar.setVisibility(View.VISIBLE);
                binding.companyName.setText("");
                binding.scoreBar.setProgress(0);
                applyStyle(context.getString(R.string.type_white));
                return;
            }

            binding.progressBar.setVisibility(View.GONE);

            companyName.setText(searchResult.name != null ? searchResult.name : searchResult.companies.get(0).name);
            applyStyle(searchResult.card_type);
            binding.companyName.setText(searchResult.name != null ? searchResult.name : searchResult.companies.get(0).name);

            if (searchResult.companies != null && searchResult.companies.get(0).plScore != null) {
                binding.scoreBar.setProgress(searchResult.companies.get(0).plScore);
            } else {
                binding.scoreBar.setProgress(0);
            }

            if (searchResult.companies != null && searchResult.companies.get(0).is_friend != null && searchResult.companies.get(
                    0).is_friend) {
                binding.heartImage.setVisibility(View.VISIBLE);
            } else {
                binding.heartImage.setVisibility(View.GONE);
            }
        }


        @Override
        public void onClick(View view) {
            int position = getAdapterPosition();
            if (position != RecyclerView.NO_POSITION) {
                onItemClicked(position);
            }
        }
    }
}

File: app/src/main/java/pl/pola_app/ui/adapter/ProductsAdapter.kt
package pl.pola_app.ui.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import butterknife.BindView
import butterknife.ButterKnife
import pl.pola_app.R
import pl.pola_app.model.SearchResult
import timber.log.Timber

class ProductsAdapter(
    private val context: Context,
    private val searchResults: ProductList
) : RecyclerView.Adapter<ProductsAdapter.ViewHolder>() {
    interface ProductClickListener {
        fun itemClicked(searchResult: SearchResult)
    }

    private var productClickListener: ProductClickListener? = null
    fun setOnProductClickListener(productClickListener: (SearchResult) -> Unit) {
        this.productClickListener = object : ProductClickListener{
            override fun itemClicked(searchResult: SearchResult) {
                productClickListener.invoke(searchResult)
            }
        }
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): ViewHolder {
        val v = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.view_product_item, viewGroup, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, i: Int) {
        val p = searchResults[i]
        viewHolder.bind(p)
    }

    override fun getItemCount(): Int {
        return searchResults.size()
    }

    private fun onItemClicked(position: Int) {
        if (productClickListener == null) {
            return
        }
        if (position > searchResults.size()) { //TODO: is it even possible to reach this state?
            Timber.w(
                IndexOutOfBoundsException(),
                "Position: %d, list size: %d",
                position,
                searchResults.size()
            )
            return
        }
        val searchResult = searchResults[position]
        searchResult?.run {
            productClickListener?.itemClicked(this)
        }
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView),
        View.OnClickListener {
        @kotlin.jvm.JvmField
        @BindView(R.id.company_name)
        var companyName: TextView? = null

        @kotlin.jvm.JvmField
        @BindView(R.id.score_bar)
        var plScore: ProgressBar? = null

        @kotlin.jvm.JvmField
        @BindView(R.id.view_product_item)
        var productCard: CardView? = null

        @kotlin.jvm.JvmField
        @BindView(R.id.progressBar)
        var progress: ProgressBar? = null

        @kotlin.jvm.JvmField
        @BindView(R.id.heart_image)
        var heartIcon: ImageView? = null
        fun bind(searchResult: SearchResult?) {
            if (searchResult == null) {
                progress?.visibility = View.VISIBLE
                companyName?.text = ""
                plScore?.progress = 0
                applyStyle(context.getString(R.string.type_white))
                return
            }
            progress?.visibility = View.GONE
            applyStyle(searchResult.card_type)
            companyName?.text = searchResult.name
            plScore?.progress = searchResult.plScore ?: 0
            if (searchResult?.is_friend == true) {
                heartIcon?.visibility = View.VISIBLE
            } else {
                heartIcon?.visibility = View.GONE
            }
        }

        private fun applyStyle(style: String) {
            val resources = context.resources
            if (style == resources.getString(R.string.type_grey)) {
                productCard?.setCardBackgroundColor(resources.getColor(R.color.card_type_grey_bk))
                plScore?.setBackgroundColor(resources.getColor(R.color.card_type_grey_score_bk))
            } else {
                productCard?.setCardBackgroundColor(resources.getColor(R.color.card_type_white_bk))
                plScore?.setBackgroundColor(resources.getColor(R.color.card_type_white_score_bk))
            }
        }

        override fun onClick(view: View) {
            val position = adapterPosition
            if (position != RecyclerView.NO_POSITION) {
                onItemClicked(position)
            }
        }

        init {
            ButterKnife.bind(this, itemView)
            itemView.setOnClickListener(this)
        }
    }
}